File: MinimalOrleansTutorial/OrleansDeviceTwins/DeviceFunction/DeviceFunctions.cs
using System.Net;
using System.Text.Json;
using DeviceFunction.HttpFunction.Models;
using Interfaces;
using Interfaces.Models;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Azure.Functions.Worker.Http;
using Microsoft.Extensions.Logging;

namespace DeviceFunction.HttpFunction;

public class DeviceFunctions(IGrainFactory factory)
{
  private readonly IGrainFactory _factory = factory ?? throw new ArgumentNullException(nameof(factory));

  [Function(nameof(UpdateDevice))]
  public async Task<HttpResponseData> UpdateDevice([HttpTrigger(AuthorizationLevel.Function, "put")] HttpRequestData req)
  {
    UpdateDeviceRequest? data = JsonSerializer.Deserialize<UpdateDeviceRequest>(req.Body);
    if(data == null) return req.CreateResponse(HttpStatusCode.BadRequest);
    
    Device newDeviceState = new (data.Connected, data.Rssi, data.Tiers.ToList());
    IDeviceGrain grain = _factory.GetGrain<IDeviceGrain>($"device@{data.DeviceId}");
    await grain.UpdateDevice(newDeviceState);
    return req.CreateResponse(HttpStatusCode.OK);
  }
}

File: MinimalOrleansTutorial/OrleansDeviceTwins/DeviceFunctionFSharp/DeviceFunctions.fs
namespace My.Function

open Microsoft.Azure.Functions.Worker
open Microsoft.Extensions.Logging
open Microsoft.Azure.Functions.Worker.Http
open System.Net

type DeviceFunctions(logger: ILogger<DeviceFunctions>) =
    
    [<Function("UpdateDevice")>]        
    member _.Run([<HttpTrigger()>]  req:HttpRequestData, executionContext:FunctionContext) =
        task {
            logger.LogInformation($"Hello at {System.DateTime.UtcNow} from an Azure function using F# on .NET 6.")
            let response = req.CreateResponse(HttpStatusCode.OK)
            do! response.WriteStringAsync("Hello World")
            return response
        }
        